### PR TITLE
Stable hashing: add comments and tests concerning platform-independence

### PR DESCRIPTION
**Pull request by tgnottingham**
SipHasher128 implements short_write in an endian-independent way, yet
its write_xxx Hasher trait methods undo this endian-independence by byte
swapping the integer inputs on big-endian hardware. StableHasher then
adds endian-independence back by also byte-swapping on big-endian
hardware prior to invoking SipHasher128.

This double swap may have the appearance of being a no-op, but is in
fact by design. In particular, we really do want SipHasher128 to be
platform-dependent, in order to be consistent with the libstd SipHasher.
Try to clarify this intent. Also, add and update a couple of unit tests.

---

Previous commit text:

~SipHasher128: fix platform-independence confusion~

~StableHasher is supposed to ensure platform independence by converting
integers to little-endian and extending isize and usize to 64 bits as
necessary, but in fact, much of that work is already handled by
SipHasher128.~

~In particular, SipHasher128 implements short_write in an
endian-independent way, yet both StableHasher and SipHasher128
additionally attempt to achieve endian-independence by byte swapping on
BE hardware before invoking short writes. This double swap has no
effect, so let's remove it.~

~Because short_write is endian-independent, SipHasher128 is already
handling part of the platform-independence, and it would be somewhat
difficult to make it *not* handle that part with the current
implementation. As splitting platform-independence responsibilities
between StableHasher and SipHasher128 would be confusing, let's make
SipHasher128 handle all of it.~

~Finally, update some incorrect comments and increase test coverage.
Unit tests pass on both LE and BE systems.~

### CODE REVIEW

**Comment by tgnottingham** [comment]
This is just a cleanup, by the way. Should be a non-functional change.

**Review by davidtwco — APPROVED** [review]
LGTM, but I'm not familiar with this code.

File: compiler/rustc_data_structures/src/stable_hasher.rs [hunk 1]
```diff
@@ -57,6 +57,9 @@ impl StableHasher {
     }
 }
 
+// SipHasher128 currently handles ensuring platform-independent results with
```
Comment (nagisa) [inline compiler/rustc_data_structures/src/stable_hasher.rs:60]: Can you add unit tests that test this property of `StableHasher` itself? We want to make sure this property holds even if we eventually move `StableHasher` away from `SipHasher`.

**Comment by nagisa** [comment]
Overall LGTM, r=me, with or without additional tests added (although if they are not added, maybe file a task for somebody else?)

File: compiler/rustc_data_structures/src/stable_hasher.rs (hunk 1 above)
Comment (tgnottingham) [reply compiler/rustc_data_structures/src/stable_hasher.rs:60]: Sure, I'll add some tests.

Comment (tgnottingham) [reply compiler/rustc_data_structures/src/stable_hasher.rs:63]: Tests added.

**Comment by tgnottingham** [comment]
Thanks @nagisa. I realize now that there might not be any process like CI running unit tests against big-endian systems, which would mean that the added tests don't yet provide the value they promise. Would the best course of action be to add a BE system to CI?

**Comment by nagisa** [comment]
It is a long standing problem that may be eventually figured out in parallel. There are some third parties (such as Debian) that might eventually run these tests, so I think the tests as they are now are good enough.

**Comment by tgnottingham** [comment]
CI failed due to what seems to be a spurious network issue. Not sure if it needs to be kicked off again.

**Comment by tgnottingham** [comment]
Saw that CI failed in #77350 on 32-bit systems due to some tests that needed updating (`test_hash_usize` in sip128/tests.rs). Pushed a change that I believe should resolve it.

**Comment by tgnottingham** [comment]
I just stumbled upon [this comment](https://github.com/rust-lang/rust/pull/68914#issuecomment-584911774), which explains that the double swapping that my PR removes was actually intentional. So I think this PR shouldn't merge after all, at least as it is now.

What I intend to do is to roll back some of the changes, but make some improvements to the comments so that it is clear why things have been done the way they've been done. @nnethercote and @michaelwoerister, please chime in if you have any input.

Thanks for the patience with the back and forth here. :)

**Comment by nnethercote** [comment]
The description says "Unit tests pass on both LE and BE systems" but subsequent comments suggest that statement might be wrong?

I was very careful in #68914 to get everything correct for LE and BE and as far as I know the current code is correct. There was a ton of back-and-forth in that PR and Michael did various tests on BE machines. The code is confusing, and I added comments to make things clearer but there is probably room for further improvements there.

(It *is* unfortunate that there is no BE testing on CI. The current state of the art is to request an account on the [GCC compile farm](https://cfarm.tetaneutral.net/) and, once you've got that, test on one of their BE machines. Horrible, I know.)

In summary: I would not be satisfied with the proposed changes in this PR unless you can show the current code does the wrong thing on BE machines, and I think you don't have that yet. But let me know if I've got that wrong.

**Comment by tgnottingham** [comment]
> The description says "Unit tests pass on both LE and BE systems" but subsequent comments suggest that statement might be wrong?

The issue wasn't LE vs BE, but 32 vs 64 bit. I had tested on x86-64 and 32-bit big-endian MIPS. However, the `test_hash_usize` functions don't have a variant that runs on the latter, so I hadn't been running tests against a 32-bit system, and the issue slipped by.

> In summary: I would not be satisfied with the proposed changes in this PR unless you can show the current code does the wrong thing on BE machines, and I think you don't have that yet. But let me know if I've got that wrong.

Oh no, the current code does the right thing on BE. I was just having a difficult time understanding the structure of the code, and thought that this would be clearer. But now I see from the other thread why things were done this way. My intent is to roll back most of my changes and just add some additional comments.

**Comment by tgnottingham** [comment]
I've made revisions and updated the PR title and description accordingly. I hope this is an improvement. :)
